File: 0140-word-break-ii/0140-word-break-ii.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        def backtrack(matched, remains):
            n = len(remains)
            for word in wordDict:
                match = copy.deepcopy(matched)
                m = len(word)
                if m <= n:
                    if word == remains[:m]:
                        match.append(word)
                        if m == n:
                            ans.append(' '.join(match))
                            return
                        else:
                            backtrack(match, remains[m:])
            return
        ans = []                     
        backtrack([], s)
        return ans

File: 0140-word-break-ii/0140-word-break-ii.cpp
class Solution {
public:
    vector<string> ans;

    vector<string> wordBreak(string s, vector<string>& wordDict) {
        sort(wordDict.begin(), wordDict.end());
        backtrack(s, wordDict, "");
        return ans;
    }
private:
    void backtrack(string s, vector<string>& wordDict, string res) {
        for (int i = 0; i < wordDict.size(); i++) {
            string word = wordDict[i];
            if (word.length() > s.length()) continue;
            if (word == s.substr(0,word.length())) {
                if (word.length() == s.length()) {
                    res += word;
                    ans.push_back(res);
                    return;
                }
                backtrack(s.substr(word.length()), wordDict, res + word + " ");
            }
        }
        return;
    }
};
//class Solution {
//public:
//    vector<string> ans;
//    vector<string> wordBreak(string s, vector<string>& wordDict) {
//        sort(wordDict.begin(), wordDict.end());
//        backtrack(s, wordDict, ""); 
//        return ans;
//    }
//    void backtrack(string s, vector<string>& wordDict, string res) {
//        for (int i = 0; i < wordDict.size(); i++) {
//            string word = wordDict[i];
//            if (word.length() > s.length()) continue;
//            if (word == s.substr(0,word.length())) {
//                if (word.length() == s.length()) {
//                    res += word;
//                    ans.push_back(res);
//                    return;
//                } else {
//                    backtrack(s.substr(word.length()), wordDict, res+word+" ");
//                }
//            }
//        }
//    }
//};